CMakeTests/CheckAndAddFlag.cmake

```diff
@@ -1,14 +1,50 @@
 include(CheckCCompilerFlag)
 include(CheckCXXCompilerFlag)
 
-macro(check_and_add_flag var flag)
-	check_c_compiler_flag(${flag} FLAG_C_${var})
-	if(FLAG_C_${var})
-		set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${flag}")
-	endif()
-
-	check_cxx_compiler_flag(${flag} FLAG_CXX_${var})
-	if(FLAG_CXX_${var})
-		set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${flag}")
-	endif()
-endmacro()
+# check_add_add_flag(<variable> <flag> [DEBUG | RELEASE])
+#
+# Add a C or C++ compilation flag to the current scope
+#
+# Can optionally add the flag to Debug or Release configurations only, use this when
+# targeting multi-configuration generators like Visual Studio or Xcode.
+# Release configurations means NOT Debug, so it will work for RelWithDebInfo or MinSizeRel too.
+#
+# If the flag is added successfully, the variables FLAG_C_<variable> and FLAG_CXX_<variable>
+# may be set to ON.
+#
+# Examples:
+#   check_and_add_flag(FOO -foo)
+#   check_and_add_flag(ONLYDEBUG -onlydebug DEBUG)
+#   check_and_add_flag(OPTMAX -O9001 RELEASE)
+
+function(check_and_add_flag var flag)
+  set(options DEBUG RELEASE)
+  set(oneValueArgs)
+  set(multiValueArgs)
+  cmake_parse_arguments(ADDFLAG "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+
+  set(compile_options)
+
+  check_c_compiler_flag(${flag} FLAG_C_${var})
+  if(FLAG_C_${var})
+    list(APPEND compile_options "$<$<COMPILE_LANGUAGE:C>:${flag}>")
+  endif()
+
+  check_cxx_compiler_flag(${flag} FLAG_CXX_${var})
+  if(FLAG_CXX_${var})
+    list(APPEND compile_options "$<$<COMPILE_LANGUAGE:CXX>:${flag}>")
+  endif()
+
+  set(config_compile_options)
+  foreach(opt ${compile_options})
+    if(ADDFLAG_DEBUG)
+      list(APPEND config_compile_options "$<$<CONFIG:Debug>:${opt}>")
```

Comment:
Instead of adding these in a later for-loop, you could add them immediately if you set up a variable (say `foo`) based on the function options, and did something like 

```cmake
add_compile_options("$<$<AND:$<COMPILE_LANGUAGE:LANG>,${foo}>:${flag}>")
```